tests/tool_excimer_sample_set_test.php

```diff
@@ -0,0 +1,167 @@
+
+    /**
+     * Tests adding samples to the object.
+     */
+    public function test_add_sample() {
+        $samples = [
+            $this->get_log_entry_stub(['a']),
+            $this->get_log_entry_stub(['b']),
+            $this->get_log_entry_stub(['c']),
+        ];
+
+        $set = new sample_set('a', 0, 1024);
+
+        $set->add_many_samples($samples);
+
+        $this->assertEquals($samples, $set->samples);
+    }
+
+    /**
+     * Tests the effect of filtering while adding samples.
+     */
+    public function test_filtering() {
+        $samples = [
+            $this->get_log_entry_stub(['a']),
+            $this->get_log_entry_stub(['b']),
+            $this->get_log_entry_stub(['c']),
+            $this->get_log_entry_stub(['d']),
+        ];
+        // This is every 2nd element of $samples.
+        $expected1 = [
+            $samples[1],
+            $samples[3]
+        ];
+        // This is every 4th element of $samples.
+        $expected2 = [
+            $samples[3]
+        ];
+
+        $set = new sample_set('a', 0, 1024);
+
+        // Each time this is called, the filter rate is doubled.
+        $set->apply_doubling();
+
+        // Filter rate should be 2, thus, only every 2nd sample should be recorded in sample set.
+        $set->add_many_samples($samples);
+
+        // Only every 2nd sample should be recorded in sample set.
+        $this->assertEquals($expected1, $set->samples);
+
+        $set = new sample_set('a', 0, 1024);
+        $set->apply_doubling();
+        $set->apply_doubling();
+
+        // Filter rate should be 4, thus, only every 4th sample should be recorded in sample set.
+        $set->add_many_samples($samples);
+
+        $this->assertEquals($expected2, $set->samples);
+    }
+
+    /**
+     * Tests stripping existing samples when calling apply_doubling.
+     */
+    public function test_stripping() {
+        $samples = [
+            $this->get_log_entry_stub(['a']),
+            $this->get_log_entry_stub(['b']),
+            $this->get_log_entry_stub(['c']),
+            $this->get_log_entry_stub(['d']),
+        ];
+        // This is $samples ofter being stripped once.
+        $expected1 = [
+            $samples[1],
+            $samples[3]
+        ];
+        // This is $samples ofter being stripped twice.
+        $expected2 = [
+            $samples[3]
+        ];
+
+        $set = new sample_set('a', 0, 1024);
+
+        $set->add_many_samples($samples);
+
+        // Every 2nd sample should be stripped after doubling.
+        $set->apply_doubling();
+        $this->assertEquals($expected1, $set->samples);
+
+        // Half of the samples should be stripped again, leaving every 4th from the original.
+        $set->apply_doubling();
+        $this->assertEquals($expected2, $set->samples);
+    }
+
+    /**
+     * Tests the invoking of apply_doubling from within add_sample.
+     */
+    public function test_doubling() {
```

Comment:
Can you change the name of this to something like `test_automatic_doubling_when_adding_samples` or similar? The test name _should_ describe what it's testing. The body of this test seems to indicate the suggested new name, whereas the current name seems more focused on testing just the apply_doubling method which is somewhat covered in other tests already.